op-service/txmgr/txmgr.go

```diff
@@ -470,44 +470,33 @@ func (m *SimpleTxManager) sendTx(ctx context.Context, tx *types.Transaction) (*t
 
 	sendState := NewSendState(m.cfg.SafeAbortNonceTooLowCount, m.cfg.TxNotInMempoolTimeout)
 	receiptChan := make(chan *types.Receipt, 1)
-	publishAndWait := func(tx *types.Transaction, bumpFees bool) *types.Transaction {
-		wg.Add(1)
-		tx, published := m.publishTx(ctx, tx, sendState, bumpFees)
-		if published {
-			go func() {
-				defer wg.Done()
-				m.waitForTx(ctx, tx, sendState, receiptChan)
-			}()
-		} else {
-			wg.Done()
-		}
-		return tx
-	}
-
-	// Immediately publish a transaction before starting the resubmission loop
-	tx = publishAndWait(tx, false)
-
 	resubmissionTimeout := m.GetBumpFeeRetryTime()
 	ticker := time.NewTicker(resubmissionTimeout)
 	defer ticker.Stop()
 
 	for {
+		if m.closed.Load() {
+			// tx manager closed while we were waiting for the tx, give up
+			m.txLogger(tx, false).Warn("TxManager closed, aborting transaction submission")
+			return nil, ErrClosed
+		}
+		if !sendState.IsWaitingForConfirmation() {
+			var published bool
+			if tx, published = m.publishTx(ctx, tx, sendState); published {
+				wg.Add(1)
+				go func() {
+					defer wg.Done()
+					m.waitForTx(ctx, tx, sendState, receiptChan)
+				}()
+			}
+		}
 		if err := sendState.CriticalError(); err != nil {
 			m.txLogger(tx, false).Warn("Aborting transaction submission", "err", err)
 			return nil, fmt.Errorf("aborted tx send due to critical error: %w", err)
 		}
+
 		select {
 		case <-ticker.C:
-			// Don't resubmit a transaction if it has been mined, but we are waiting for the conf depth.
-			if sendState.IsWaitingForConfirmation() {
-				continue
-			}
-			// if the tx manager closed while we were waiting for the tx, give up
-			if m.closed.Load() {
-				m.txLogger(tx, false).Warn("TxManager closed, aborting transaction submission")
-				return nil, ErrClosed
-			}
```

Comment:
I vaguely remember that the ordering here was on purpose, to wait for confirmation even if the txmgr got closed, so that any in-flight txs will be waited for before shutting down. It looks like you reordered this above. Or does it actually still work because we `defer wg.Wait()` at the start of `sendTx`?

Reply:
changed to wait for any pending-for-confirmation transactions before shutting down.

Reply:
Oh yes, we cancel the context.

Yes the idea was that an orderly txmgr shutdown should still confirm in-flight transactions, so that the next time the component that is using the txmgr starts up, its latest tx got confirmed on-chain. The idea is that this component will likely read some on-chain state on startup to determine its next operation, so we avoid duplicate txs.

Reply:
got it, ok should work as intended now then.